### PR TITLE
feat(npm): Basic support for `cap` update strategy [WIP]

### LINKED ISSUE #3861: rangeStrategy=cap
**What would you like Renovate to be able to do?**

Support a new `rangeStrategy` called "cap" whose purpose is retain ranges but to make sure they have a "cap" (ceiling/maximum).

**Describe the solution you'd like**

Renovate should determine if the current range is capped or uncapped, and cap it if necessary. For example if the current range is `^1.0.0` then it can be capped by converting that range to `^1.0.0 <=1.2.0` (assuming `1.2.0` is the latest version).

One way to approach this would be these steps:
- If the currentValue is a range, find the highest existing version that satisfies it
- Increment that highest existing version by the minimum possible (e.g. patch) and test if the incremented version also satisfies the current range
- If it satisfies it, it means we need to cap and can add `<=` or equivalent to the range
- Double check make sure the incremented version doesn't satisfy the new capped range. If it does, we want to raise a warning because we have an edge case.

I think we can start with `npm`, but `maven` and `ruby` versioning are also important.

### PR DESCRIPTION
**Pull request by zharinov**
Closes #3861

### CODE REVIEW

File: test/versioning/npm.spec.js
Comment (zharinov) [inline test/versioning/npm.spec.js]: @rarkins  please review these examples out

**Comment by zharinov** [comment]
@rarkins Seems I'm done here

**Comment by rarkins** [comment]
@zharinov I just realised a good idea for capping which probably impacts your work here. I think that "capping" a range should be logically similar to "pinning" it. i.e. we do not upgrade a range at the same time as we're pinning it, and we hold back on upgrades to ranges until after there is a pin PR merged for it. I think we should do the same for capping. i.e. if a user opts in to capping then they would get a "Cap dependencies" PR instead of "Pin dependencies", etc.

Does this make sense to you and does it require any changes to the version range? Please look at how we implement `'pin'` and follow that approach.

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: I don't think that `1.0.7-prerelease.1` falls within the range `^1` so this may not be a valid use case.

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: Change to `~>1.1.0 <= 1.1.7`

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: How can the from value be `1.0.0` if the range is `1.2.*` ?

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: I think `1.0.0 || 1.1.0 || 1.2.3`. i.e if something is already capped then we should default to our regular behaviour

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: `~1.1 <=1.1.7`

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: Same as earlier. Is it valid?

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: `^2 <=2.1.7`

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: `5.1 <= 5.1.7`

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: Is this a duplicate with line above?

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: `1.1.* <= 1.1.0`

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: `2.* <=2.1.0`

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: I don't think that `0.7.0-rc.2` falls within range `~0.6.1` so this is invalid. I think the answer needs to be something like `~0.7.0-rc <=0.7.0-rc.2`

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: This is nearly impossible to guess right, and unlikely to happen in the wild. I almost feel like skipping it. Technically it would need to be something like: `(5.0 <=5.0.2) || (5.5 <= 5.5.4) || (6.1 <=6.1.7)` and I don't even know if that's valid.

**Review by rarkins — CHANGES_REQUESTED** [review]
Having a chance to review the examples it's now clearer to me how we should do this:
1. If a range needs capping, then we should use the same logic we do when a range needs pinning.
2. To determine what the final range should be, we should use this logic:
- First, apply our `replace` logic to the main range
- Second, add the cap
3. I think for now we skip complex ranges and simply extend them instead of capping

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: Why is this different from the two below it?

File: test/versioning/npm.spec.js
Comment (rarkins) [inline test/versioning/npm.spec.js]: There's a bit of inconsistency here

**Review by rarkins — CHANGES_REQUESTED** [review]
I'm not 100% sure the best way to approach this.

If we break it into two phases like pinning, then we should not need to consider cases like `['1.*', '2.3.4', '>=1.0.0 <=2.3.4'], // '>=1.0.0 <2.0.0'` and it's much simpler.

i.e. 

step 1: cap existing range with the highest value that matches it
step 2: widen capped ranges as newer versions are released

BTW if the current range is `1.* <=1.2.0` and then version `2.0.0` is released then our widening result should be something like `1.* <= 1.2.0 || 2.* <= 2.0.0`.

**Comment by rarkins** [comment]
Let's implement it like "pin":
- First step is to "cap" existing ranges using `<= x`
- Other updates to a dependency are blocked by the "cap" PR
- Once capped, ranges will be "complex" and we will use our existing "widen" approach whenever new releases come out

In this case we don't really need to worry about getting ones like `['1.x', '2.3.4', '>=1.0.0 <=2.3.4'], // '>=1.0.0 <2.0.0'` perfect.

If the range is capped, then we widen it. If it's uncapped, then just call "replace" as a temporary measure.

How to tell if a range is uncapped?
- Calculate the max satisfying version
- Increment that version using a pre-patch
- Check if it still satisfies

**Comment by zharinov** [comment]
> First step is to "cap" existing ranges using <= x

Where `x` is the max satisfying of existing range?
And if it's not a range: values like "1.2.3" become ranges like "<= 1.2.3".

> How to tell if a range is uncapped?
> 
> * Calculate the max satisfying version
> * Increment that version using a pre-patch
> * Check if it still satisfies

I'm not sure how it works, let's elaborate on that. From some step I can't get the idea:
```
1. allVersions    = [0.9, 1.0, 1.1, 1.2, 2.0]
2. currentValue   = 1.*
3. maxSatisfying  = 1.2
4. prePatch       = 1.2-0 (and of course 1.2-0 < 1.2)
5. patchSatisfies = 1.1   (max satisfying of "<= 1.2-0" range)
6. 1.1 != 1.2     => range is capped or uncapped?
```

**Comment by rarkins** [comment]
> And if it's not a range: values like "1.2.3" become ranges like "<= 1.2.3".

If it's a versioning scheme where `1.2.3` means "exactly 1.2.3" then we don't need to convert it. We only cap ranges if they are ranges to begin with.

BTW I'm not sure why I wrote "pre-patch". Forget that for now, let's try again..

I think we maybe could utilize a new optional function in each versionScheme that does "minimum increment".

We could start by keeping it simple e.g increment a patch for `semver`. So `semver.nextVersion("1.2.3") === "1.2.4"`. Eventually we could also handle the edge case of pre-releases better, e.g. `semver.nextVersion("1.2.3-alpha.2") === "1.2.3-alpha.3"`

What do you think?

**Comment by zharinov** [comment]
Hm, still not obvious to me. By the way, `cap`-style ranges seem to have well-defined pattern (e.g. `<= X.X.X` and `(,X.X.X)`) and functionally are more similar to versions instead of ranges. What if (1) convert it to a regular version, (2) replace it and then (3) convert back? Would couple functions with nullable defaults in versioning api helping here?

**Comment by rarkins** [comment]
We want to know "is this existing range capped?". And if the answer is no, then limit it to be no greater than the highest matching version already.

Naturally we can see if `versionScheme.isSingleVersion(currentValue) === true`, because if so then obviously it's not a range so must be capped.

It doesn't help us with determining if a range is capped or not. Instead of trying to apply some type of "reasoning", my idea is just to do this:
- Find the highest existing value that satisfies the range
- Increment that value by the minimum amount possible
- Test if that new value still satisfies the range

If the above is true then we cap it by adding `<= ${highestCurrentMatchingVersion}` to the range (or equivalent).

The missing piece is "Increment that value by the minimum amount possible" so that's why I was thinking of adding a function to our version schemes called "increment" or "next version" that returns the smallest next version. Then the version schemes only require one function added, and it's a pretty small function.

**Comment by zharinov** [comment]
Okay, here I have examples produced by this [minimal implementation](https://github.com/renovatebot/renovate/pull/4923/files): for [x-ranges](https://github.com/zharinov/renovate-npm-minimal-1/pulls?q=is%3Apr+is%3Aclosed), [caret-ranges](https://github.com/zharinov/renovate-npm-minimal-2/pulls?q=is%3Apr+is%3Aclosed)  and [for single version](https://github.com/zharinov/renovate-npm-minimal-3/pulls?q=is%3Apr+is%3Aclosed).
I emulated new version by downgrading the value after initial `cap` PR merge.

**Comment by zharinov** [comment]
If [this way for range extending](https://github.com/renovatebot/renovate/pull/4923/files#diff-7dbd9703d5a7714bf7631d123bca37fbR24) is good enough, then we can close this PR I think

**Pull request closed without merging**
